File: backend/gmail.py
import base64
import os.path
from email.utils import parsedate_to_datetime
from typing import Any, Dict

from google.auth.exceptions import RefreshError
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build

from .config import MAX_BODY_CHARS, SCOPES

TOKEN_PATH = "token.json"
CREDENTIALS_PATH = "credentials.json"

def _header_map(headers):
    mapped = {}
    for header in headers:
        name = header.get("name", "").lower()
        if name and name not in mapped:
            mapped[name] = header.get("value", "")
    return mapped


def iter_parts(payload):
    stack = [payload]
    while stack:
        part = stack.pop()
        yield part
        for subpart in part.get("parts", []) or []:
            stack.append(subpart)


def decode_body(data):
    if not data:
        return ""
    return base64.urlsafe_b64decode(data.encode("utf-8")).decode("utf-8", errors="replace")


def _body_and_attachments(payload):
    plain_chunks = []
    html_chunks = []
    attachments = False
    for part in iter_parts(payload):
        body = part.get("body", {}) or {}
        if part.get("filename") or body.get("attachmentId"):
            attachments = True
        body_data = body.get("data")
        if not body_data:
            continue
        decoded = decode_body(body_data)
        mime_type = part.get("mimeType", "")
        if mime_type == "text/plain":
            plain_chunks.append(decoded)
        elif mime_type == "text/html":
            html_chunks.append(decoded)
    body = "\n\n".join(plain_chunks or html_chunks)
    return body, attachments


def _truncate(text: str, limit: int) -> str:
    if len(text) <= limit:
        return text
    return f"{text[:limit]}\n\n[truncated]"


def get_details(message):
    payload = message.get("payload", {})
    headers = _header_map(payload.get("headers", []))
    subject = headers.get("subject") or "No subject"
    sender = headers.get("from") or "Unknown sender"
    date_header = headers.get("date")
    date_time = ""
    if date_header:
        parsed = parsedate_to_datetime(date_header)
        if parsed:
            date_time = parsed.isoformat()
    snippet = message.get("snippet", "")
    body, attachments = _body_and_attachments(payload)
    return {
        "subject": subject,
        "sender": sender,
        "date_time": date_time or date_header or "Unknown date",
        "snippet": snippet,
        "attachments": attachments,
        "body": body,
    }


def normalize_email_details(details: Dict[str, Any]) -> Dict[str, Any]:
    body = details.get("body", "") or ""
    return {
        "id": details.get("message_id") or details.get("id") or "",
        "subject": details.get("subject", ""),
        "sender": details.get("sender", ""),
        "date_time": details.get("date_time", ""),
        "snippet": details.get("snippet", ""),
        "attachments": bool(details.get("attachments")),
        "body": _truncate(body, MAX_BODY_CHARS),
    }


def get_gmail_service(scopes=None):
    scopes = scopes or SCOPES
    creds = None
    # The file token.json stores the user's access and refresh tokens, and is
    # created automatically when the authorization flow completes for the first
    # time.
    if os.path.exists(TOKEN_PATH):
        creds = Credentials.from_authorized_user_file(TOKEN_PATH, scopes)

    def _run_oauth_flow():
        flow = InstalledAppFlow.from_client_secrets_file(CREDENTIALS_PATH, scopes)
        return flow.run_local_server(port=0)

    # If there are no (valid) credentials available, let the user log in.
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            try:
                creds.refresh(Request())
            except RefreshError:
                # Token can become invalid/revoked; remove stale token and re-auth.
                if os.path.exists(TOKEN_PATH):
                    os.remove(TOKEN_PATH)
                creds = _run_oauth_flow()
        else:
            creds = _run_oauth_flow()
        # Save the credentials for the next run
        with open(TOKEN_PATH, "w", encoding="utf-8") as token:
            token.write(creds.to_json())
    return build("gmail", "v1", credentials=creds)


def fetch_unread_message_details(max_results=2, user_id="me", label_ids=None):
    service = get_gmail_service()
    label_ids = label_ids or ["INBOX", "UNREAD"]
    results = (
        service.users()
        .messages()
        .list(userId=user_id, labelIds=label_ids, maxResults=max_results)
        .execute()
    )
    messages = results.get("messages", [])
    details_list = []
    for message in messages:
        msg = service.users().messages().get(userId=user_id, id=message["id"]).execute()
        details = get_details(msg)
        details["message_id"] = message["id"]
        details_list.append(details)
    return details_list

File: web/lib/backend/gmail.ts
import fs from "node:fs";

import {
  CREDENTIALS_PATH,
  MAX_BODY_CHARS,
  SCOPES,
  TOKEN_PATH,
} from "@/lib/backend/config";
import type {
  EmailDetails,
  NormalizedEmailDetails,
  UnknownRecord,
} from "@/lib/backend/types";

type GoogleInstalledClient = {
  client_id: string;
  client_secret: string;
  token_uri?: string;
};

type GoogleCredentialsFile = {
  installed?: GoogleInstalledClient;
  web?: GoogleInstalledClient;
};

type StoredToken = {
  token?: string;
  refresh_token?: string;
  token_uri?: string;
  client_id?: string;
  client_secret?: string;
  expiry?: string;
  scopes?: string[];
};

type GmailListResponse = {
  messages?: Array<{ id: string }>;
};

type GmailHeader = {
  name?: string;
  value?: string;
};

type GmailPart = {
  mimeType?: string;
  filename?: string;
  body?: {
    data?: string;
    attachmentId?: string;
  };
  parts?: GmailPart[];
  headers?: GmailHeader[];
};

type GmailMessage = {
  id?: string;
  snippet?: string;
  payload?: GmailPart;
};

function isRecord(value: unknown): value is UnknownRecord {
  return typeof value === "object" && value !== null;
}

function readJsonFile(path: string): unknown {
  return JSON.parse(fs.readFileSync(path, "utf-8"));
}

function loadGoogleClient(): GoogleInstalledClient {
  if (!fs.existsSync(CREDENTIALS_PATH)) {
    throw new Error(`Missing credentials file at ${CREDENTIALS_PATH}`);
  }

  const parsed = readJsonFile(CREDENTIALS_PATH) as GoogleCredentialsFile;
  const client = parsed.installed ?? parsed.web;
  if (!client?.client_id || !client.client_secret) {
    throw new Error("Invalid credentials.json format.");
  }
  return client;
}

function loadStoredToken(): StoredToken | null {
  if (!fs.existsSync(TOKEN_PATH)) {
    return null;
  }
  const parsed = readJsonFile(TOKEN_PATH);
  return isRecord(parsed) ? (parsed as StoredToken) : null;
}

function saveStoredToken(token: StoredToken): void {
  fs.writeFileSync(TOKEN_PATH, `${JSON.stringify(token, null, 2)}\n`, "utf-8");
}

function hasUsableToken(token: StoredToken): boolean {
  if (!token.token) {
    return false;
  }
  if (!token.expiry) {
    return true;
  }
  const expiryMs = Date.parse(token.expiry);
  if (Number.isNaN(expiryMs)) {
    return true;
  }
  return expiryMs > Date.now() + 60_000;
}

async function refreshToken(token: StoredToken): Promise<StoredToken> {
  const client = loadGoogleClient();
  const refreshTokenValue = token.refresh_token;
  if (!refreshTokenValue) {
    throw new Error(
      `Missing refresh token in ${TOKEN_PATH}. Recreate token.json via OAuth once.`,
    );
  }

  const tokenUri = token.token_uri ?? client.token_uri ?? "https://oauth2.googleapis.com/token";
  const body = new URLSearchParams({
    client_id: client.client_id,
    client_secret: client.client_secret,
    refresh_token: refreshTokenValue,
    grant_type: "refresh_token",
  });

  const response = await fetch(tokenUri, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: body.toString(),
  });

  const payload = (await response.json().catch(() => ({}))) as UnknownRecord;
  if (!response.ok) {
    throw new Error(
      `Google token refresh failed (${response.status}): ${JSON.stringify(payload)}`,
    );
  }

  const accessToken =
    typeof payload.access_token === "string" ? payload.access_token : undefined;
  const expiresIn =
    typeof payload.expires_in === "number" ? payload.expires_in : undefined;
  const refreshedToken =
    typeof payload.refresh_token === "string"
      ? payload.refresh_token
      : refreshTokenValue;

  if (!accessToken) {
    throw new Error("Google refresh response missing access_token.");
  }

  const nextToken: StoredToken = {
    ...token,
    token: accessToken,
    refresh_token: refreshedToken,
    token_uri: tokenUri,
    client_id: client.client_id,
    client_secret: client.client_secret,
    scopes: token.scopes ?? SCOPES,
  };

  if (expiresIn) {
    nextToken.expiry = new Date(Date.now() + expiresIn * 1000).toISOString();
  }

  saveStoredToken(nextToken);
  return nextToken;
}

async function getAccessToken(): Promise<string> {
  const token = loadStoredToken();
  if (!token) {
    throw new Error(
      `Missing ${TOKEN_PATH}. Run OAuth bootstrap once to generate token.json.`,
    );
  }
  if (hasUsableToken(token)) {
    return token.token as string;
  }
  const refreshed = await refreshToken(token);
  if (!refreshed.token) {
    throw new Error("Unable to obtain Gmail access token.");
  }
  return refreshed.token;
}

async function gmailGet<T>(
  apiPath: string,
  accessToken: string,
  params: URLSearchParams,
): Promise<T> {
  const url = `https://gmail.googleapis.com/gmail/v1/${apiPath}?${params.toString()}`;
  const response = await fetch(url, {
    method: "GET",
    headers: {
      Authorization: `Bearer ${accessToken}`,
    },
    cache: "no-store",
  });

  const contentType = response.headers.get("content-type") ?? "";
  const body = contentType.includes("application/json")
    ? await response.json()
    : await response.text();
  if (!response.ok) {
    throw new Error(
      `Gmail API ${apiPath} failed (${response.status}): ${
        typeof body === "string" ? body : JSON.stringify(body)
      }`,
    );
  }
  return body as T;
}

function headerMap(headers: GmailHeader[] | undefined): Record<string, string> {
  const mapped: Record<string, string> = {};
  if (!headers) {
    return mapped;
  }

  for (const header of headers) {
    const name =
      typeof header.name === "string" ? header.name.toLowerCase() : undefined;
    if (!name || mapped[name]) {
      continue;
    }
    mapped[name] = typeof header.value === "string" ? header.value : "";
  }
  return mapped;
}

function decodeBody(data: string | undefined): string {
  if (!data) {
    return "";
  }

  const normalized = data.replace(/-/g, "+").replace(/_/g, "/");
  const padding = normalized.length % 4 === 0 ? "" : "=".repeat(4 - (normalized.length % 4));
  return Buffer.from(`${normalized}${padding}`, "base64").toString("utf-8");
}

function* iterParts(payload: GmailPart | undefined): Generator<GmailPart> {
  if (!payload) {
    return;
  }
  const stack: GmailPart[] = [payload];
  while (stack.length > 0) {
    const current = stack.pop() as GmailPart;
    yield current;
    for (const part of current.parts ?? []) {
      stack.push(part);
    }
  }
}

function bodyAndAttachments(payload: GmailPart | undefined): {
  body: string;
  attachments: boolean;
} {
  const plainChunks: string[] = [];
  const htmlChunks: string[] = [];
  let attachments = false;

  for (const part of iterParts(payload)) {
    const body = part.body ?? {};
    if (part.filename || body.attachmentId) {
      attachments = true;
    }
    if (!body.data) {
      continue;
    }
    const decoded = decodeBody(body.data);
    if (part.mimeType === "text/plain") {
      plainChunks.push(decoded);
    } else if (part.mimeType === "text/html") {
      htmlChunks.push(decoded);
    }
  }

  return {
    body: (plainChunks.length > 0 ? plainChunks : htmlChunks).join("\n\n"),
    attachments,
  };
}

function truncate(text: string, limit: number): string {
  if (text.length <= limit) {
    return text;
  }
  return `${text.slice(0, limit)}\n\n[truncated]`;
}

function normalizeDate(dateHeader: string | undefined): string {
  if (!dateHeader) {
    return "Unknown date";
  }
  const parsed = new Date(dateHeader);
  if (Number.isNaN(parsed.getTime())) {
    return dateHeader;
  }
  return parsed.toISOString();
}

function getDetails(message: GmailMessage): EmailDetails {
  const payload = message.payload;
  const headers = headerMap(payload?.headers);
  const parsed = bodyAndAttachments(payload);

  return {
    subject: headers.subject || "No subject",
    sender: headers.from || "Unknown sender",
    date_time: normalizeDate(headers.date),
    snippet: typeof message.snippet === "string" ? message.snippet : "",
    attachments: parsed.attachments,
    body: parsed.body,
  };
}

export function normalizeEmailDetails(details: EmailDetails): NormalizedEmailDetails {
  const body = details.body ?? "";
  return {
    id: details.message_id ?? details.id ?? "",
    subject: details.subject ?? "",
    sender: details.sender ?? "",
    date_time: details.date_time ?? "",
    snippet: details.snippet ?? "",
    attachments: Boolean(details.attachments),
    body: truncate(body, MAX_BODY_CHARS),
  };
}

export async function fetchUnreadMessageDetails(
  maxResults: number = 2,
  userId: string = "me",
  labelIds: string[] = ["INBOX", "UNREAD"],
): Promise<EmailDetails[]> {
  const accessToken = await getAccessToken();
  const listParams = new URLSearchParams({ maxResults: String(maxResults) });
  for (const labelId of labelIds) {
    listParams.append("labelIds", labelId);
  }

  const list = await gmailGet<GmailListResponse>(
    `users/${encodeURIComponent(userId)}/messages`,
    accessToken,
    listParams,
  );
  const messages = Array.isArray(list.messages) ? list.messages : [];

  const details = await Promise.all(
    messages.map(async (message) => {
      const getParams = new URLSearchParams({ format: "full" });
      const full = await gmailGet<GmailMessage>(
        `users/${encodeURIComponent(userId)}/messages/${encodeURIComponent(message.id)}`,
        accessToken,
        getParams,
      );
      const item = getDetails(full);
      item.message_id = message.id;
      return item;
    }),
  );

  return details;
}
